plutus-use-cases/src/Plutus/Contracts/Uniswap.hs

```diff
@@ -0,0 +1,988 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveAnyClass             #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE NoImplicitPrelude          #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RecordWildCards            #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TemplateHaskell            #-}
+{-# LANGUAGE TypeApplications           #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE TypeOperators              #-}
+
+-- | A decentralized exchange for arbitrary token pairs following the
+-- [Uniswap protocol](https://uniswap.org/whitepaper.pdf).
+--
+module Plutus.Contracts.Uniswap
+    ( Coin (..)
+    , coin, coinValueOf
+    , Uniswap (..), uniswap
+    , poolStateCoinFromUniswapCurrency, liquidityCoin
+    , CreateParams (..)
+    , SwapParams (..)
+    , CloseParams (..)
+    , RemoveParams (..)
+    , AddParams (..)
+    , UniswapUserSchema, UserContractState (..)
+    , UniswapOwnerSchema
+    , start, create, add, remove, close, swap, pools
+    , ownerEndpoint, userEndpoints
+    ) where
+
+import           Control.Monad                    hiding (fmap)
+import qualified Data.Map                         as Map
+import           Data.Monoid                      (Last (..))
+import           Data.Proxy                       (Proxy (..))
+import           Data.Text                        (Text, pack)
+import           Data.Void                        (Void)
+import           Ledger                           hiding (singleton)
+import           Ledger.Constraints               as Constraints
+import           Ledger.Constraints.OnChain       as Constraints
+import           Ledger.Constraints.TxConstraints as Constraints
+import qualified Ledger.Typed.Scripts             as Scripts
+import           Ledger.Value                     as Value
+import           Playground.Contract
+import           Plutus.Contract                  hiding (when)
+import qualified Plutus.Contracts.Currency        as Currency
+import qualified PlutusTx
+import           PlutusTx.Prelude                 hiding (Semigroup (..), unless)
+import           Prelude                          (Semigroup (..))
+import qualified Prelude
+import           Text.Printf                      (printf)
+
+feeNum, feeDen :: Integer
```

Comment:
haddocks

Reply:
It would be nice to have a note explaining at a high level how this is implemented, stuff like what all the tokens are and what they control, etc.